### DISCUSSION TITLE
Is Mezzanine still viable?

### DISCUSSION THREAD

**Discussion by rich-hohm** [stephenmcd/mezzanine#2038] [General]
I worked with Mezzanine several years ago, and liked what I saw, a lot. This past week I reviewed it for possible use in a hobbyist club website I'm rebuilding. Here's what I found:

- Frequent and recent updates - good, even though 6.0.0 and almost all the 5.x releases have been one or two misc. changes or bug fixes each. 
- The project website appears dead. mezzanine.jupo.org redirects here, and all other links I tried yielded 404.
- The only documentation I found (on readthedocs.io) is for version 4.3.1. This turns out to be less of a problem than it appeared since, as noted above, there appear to be very few functional changes since 4.3.x, but some sections appear to be seriously out of date.
- With the exception of Cartridge and related packages, every plugin I looked at was stale or dead. None supported Mez > 3.x or Django > 2.x.
- No premium themes available. None. The last time I worked with Mez, RocketTheme and most others had at least a modest selection of Mez themes available. I purchased one to customize it for the site I was building at the time. Now, nothing.
- None of the free themes I discovered claimed to support Mez > 4.x, and none that I tried in my test installation worked. I have a link to a theme-building tutorial, but it looks to be rather effort-intensive, much more so than customizing an existing theme.
- The "enthusiastic, vibrant user community" the various review sites applaud appears to be, well, here, with all of a half-dozen topics opened in the past six months. StackOverflow has had four Mez threads in the past _three years_, and it looks like two of those actually involved Django issues.

So. I have a site that is currently running django-cms and django-cms-blog, which turn out to be remarkably buggy and user-unfriendly for such popular packages, and I want to replace them. My question is, even though I like Mezzanine and it does 80% of what this site needs to do right out of the box, it looks like it's dying. Is there any point in even considering it over other alternatives like wagtail (which would need a custom blog; puput isn't up to the job), or abandoning django altogether in favor of something like Plone or Drupal or Joomla?

**Comment by kenbolton** [comment]
> Frequent and recent updates - good, even though 6.0.0 and almost all the 5.x releases have been one or two misc. changes or bug fixes each.

To me, this speaks to stability. Do not discount the hard work required to support a fairly broad range of Python and Django versions. We have all seen the dumpster fire of "move fast and break s4!7".

> The project website appears dead. mezzanine.jupo.org redirects here, and all other links I tried yielded 404.

The owner of jupo.org is no longer actively working on the project. There are a couple of issues in the tracker discussing this. The README is the current stand-in for https://github.com/stephenmcd/mezzanine.jupo.org

> The only documentation I found (on readthedocs.io) is for version 4.3.1. This turns out to be less of a problem than it appeared since, as noted above, there appear to be very few functional changes since 4.3.x, but some sections appear to be seriously out of date.

Here you go! https://github.com/stephenmcd/mezzanine/tree/master/docs

> With the exception of Cartridge and related packages, every plugin I looked at was stale or dead. None supported Mez > 3.x or Django > 2.x.

Examples? My own tax plugin for cartridge is open source and could be upgraded, but I have more pressing tasks. 

> No premium themes available. None. The last time I worked with Mez, RocketTheme and most others had at least a modest selection of Mez themes available. I purchased one to customize it for the site I was building at the time. Now, nothing.

Willing to pay for a theme? Hire someone to update it. But as you pointed out above, "there have been one or two  misc .changes or bug fixes", and nothing in the last year (or longer) has touched the CSS and HTML enough to break themes. I suspect an hour or two of diligent use of a web browser's developer tools will resolve most of the issues, then iterate as you discover other parts that need change.

 > None of the free themes I discovered claimed to support Mez > 4.x, and none that I tried in my test installation worked. I have a link to a theme-building tutorial, but it looks to be rather effort-intensive, much more so than customizing an existing theme.

Ain't no free lunches, friend.

> The "enthusiastic, vibrant user community" the various review sites applaud appears to be, well, here, with all of a half-dozen topics opened in the past six months. StackOverflow has had four Mez threads in the past three years, and it looks like two of those actually involved Django issues.

We're here! (Where else would we be?) There are few issues opened because "it just works". SO is not where any community I want to be a part of is spending time.


I came to Mezzanine as a refugee from django-cms, and before that Drupal, and before that Plone. The only one that worked well and was pleasant to work with was Plone, but Plone's dependency on Zope was often a dealbreaker for my clients. django-cms was always a nightmare, and Drupal isn't Python. Wagtail is pretty cool, though they never acknowledged the fact that it is derived from Mezzanine.

**Reply by kenbolton** [reply]
The answer to your question is "Yes". Mezzanine is absolutely viable. It remains a pleasure to work with and on Mezzanine. But it will be work. It will require reading the docs and the source.  There are no open source projects that come to mind for which that is not the case.

**Comment by molokov** [comment]
I do hope Mezzanine is still viable, I'm still using it! My website is a few versions behind (Django 2.x) but I do plan to update soon, as we're about to enter a quiet period between major events. Yes, the original big push done by Stephen may have faded because he's not able to be involved any more, but Ken, Henri and others have done amazing work keeping this project alive. I only wish I had time to contribute more, but I am truly grateful for everything done by those do contribute upgrades to Mezzanine and related projects, making sure it works with the latest Django versions.

**Reply by hamzabawumia** [reply]
please what version of Mezzanine are you using with the Django 2.2 it seem there are serious compartibility issues between versions of python, django and mezzanine. 

I installed newest version mezzanine==6.0.0, django==4.x and python 3.11

In get errors in the admin when ever i try to save a record:

TypeError at /admin/pages/richtextpage/1/change/</h1>
TypeError unsupported operand type(s) for +: 'frozenset' and 'list'

I have tried all permutations of versions but each one gives a different compartibility error

**Reply by Parthian** [reply]
Here are my notes on this problem. An easy fix. I can't recall the bit about Bleach but I've kept it in for completeness.

_Bleach, which is probably all but deprecated, had a bug with frozenset and list - crashed silently, needed settings debug=true

....python3.10/site-packages/mezzanine/utils/html.py
        # protocols=ALLOWED_PROTOCOLS + ["tel"],    Needs a list so replace with -
        protocols=list(ALLOWED_PROTOCOLS) + ["tel"],_

 See https://blog.lordvan.com/blog/upgrading-mezzanine-forced-to-due-to-move-to-python310/

**Reply by molokov** [reply]
Dear @hamzabawumia 

The issue you describe is reported here: https://github.com/stephenmcd/mezzanine/issues/2054 and the solution, as mentioned, is to use an earlier bleach version (bleach < 6) or to apply this fix: https://github.com/stephenmcd/mezzanine/pull/2059/files

I have Mezzanine working well with with the latest Django (5.1.x) and have only had to make a few minor changes to get it to work this way. Feel free to use my fork at https://github.com/molokov/mezzanine or to create your own to apply some fixes.
(I should probably also update some of it specifically to test on Django 5.1, but it does still work).

I'm not sure if the maintenance team for the main repo has time to work on this any more, but that doesn't mean you can't get the project to work with only some minor fixes of your own.

**Comment by rapto** [comment]
We use it extensively, but still have to migrate from Django 2.2. I'd
rather make the migration and correct whatever I find than jump onto
another cms. I have found very few quirks in the software.

**Reply by hamzabawumia** [reply]
@kenbolton please can you give us a version compartibility list. It seems when you install it works out of the box but when you start editting stuff in the admin you start getting errors. like so:

TypeError at /admin/pages/richtextpage/1/change/
TypeError unsupported operand type(s) for +: 'frozenset' and 'list'

YOU HAVE A GREAT PRODUCT but its a shame that you are not testing it and fixing python and django version compartibilities. It always looks great on launch but when you start using the admin, then you see several errors showing up.
